// salesforce-alm/dist/commands/force/mdapi/describemetadata.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdapiDescribemetadataCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const Messages = require("../../../lib/messages");
const messages = Messages();
class MdapiDescribemetadataCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        // eslint-disable-next-line @typescript-eslint/no-shadow
        const MdapiDescribemetadataCommandImpl = require('../../../lib/mdapi/mdapiDescribemetadataCommand');
        return this.execLegacyCommand(new MdapiDescribemetadataCommandImpl(context), context);
    }
}
exports.MdapiDescribemetadataCommand = MdapiDescribemetadataCommand;
MdapiDescribemetadataCommand.theDescription = messages.getMessage('mdDescribeMetadataCommandCliDescription');
MdapiDescribemetadataCommand.longDescription = messages.getMessage('mdDescribeMetadataCommandCliLong');
MdapiDescribemetadataCommand.help = messages.getMessage('mdDescribeMetadataCommandCliHelp');
MdapiDescribemetadataCommand.requiresProject = false;
MdapiDescribemetadataCommand.requiresUsername = true;
MdapiDescribemetadataCommand.flagsConfig = {
    apiversion: command_1.flags.builtin({
        // @ts-ignore force char override for backward compat -- don't try this at home!
        char: 'a',
        description: messages.getMessage('mdDescribeMetadataCommandCliApiVersion'),
        longDescription: messages.getMessage('mdDescribeMetadataCommandCliApiVersionLong'),
    }),
    resultfile: command_1.flags.filepath({
        char: 'f',
        description: messages.getMessage('mdDescribeMetadataCommandCliFilterKnown'),
        longDescription: messages.getMessage('mdDescribeMetadataCommandCliFilterKnownLong'),
        required: false,
    }),
    filterknown: command_1.flags.boolean({
        char: 'k',
        description: messages.getMessage('mdDescribeMetadataCommandCliResultFile'),
        longDescription: messages.getMessage('mdDescribeMetadataCommandCliResultFileLong'),
        hidden: true,
    }),
};
