// @salesforce/plugin-source/lib/formatters/deployProgressBarFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployProgressBarFormatter = void 0;
const kit_1 = require("@salesforce/kit");
const cli_ux_1 = require("cli-ux");
const progressFormatter_1 = require("./progressFormatter");
class DeployProgressBarFormatter extends progressFormatter_1.ProgressFormatter {
    constructor(logger, ux) {
        super(logger, ux);
    }
    // displays the progress of the Deployment
    progress(deploy) {
        this.initProgressBar();
        const startProgressBar = (0, kit_1.once)((componentTotal) => {
            this.progressBar.start(componentTotal);
        });
        deploy.onUpdate((data) => {
            // the numCompTot. isn't computed right away, wait to start until we know how many we have
            if (data.numberComponentsTotal) {
                startProgressBar(data.numberComponentsTotal + data.numberTestsTotal);
                this.progressBar.update(data.numberComponentsDeployed + data.numberTestsCompleted);
            }
            // the numTestsTot. isn't computed until validated as tests by the server, update the PB once we know
            if (data.numberTestsTotal && data.numberComponentsTotal) {
                this.progressBar.setTotal(data.numberComponentsTotal + data.numberTestsTotal);
            }
        });
        // any thing else should stop the progress bar
        deploy.onFinish((data) => {
            // the final tick of `onUpdate` is actually fired with `onFinish`
            this.progressBar.update(data.response.numberComponentsDeployed + data.response.numberTestsCompleted);
            this.progressBar.stop();
        });
        deploy.onCancel(() => {
            this.progressBar.stop();
        });
        deploy.onError((error) => {
            this.progressBar.stop();
            throw error;
        });
    }
    // used to intialise the progress bar
    initProgressBar() {
        this.logger.debug('initializing progress bar');
        this.progressBar = cli_ux_1.default.progress({
            format: 'SOURCE PROGRESS | {bar} | {value}/{total} Components',
            barCompleteChar: '\u2588',
            barIncompleteChar: '\u2591',
            linewrap: true,
        });
    }
}
exports.DeployProgressBarFormatter = DeployProgressBarFormatter;
