// @salesforce/sf-plugins-core/lib/deployer.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deployer = exports.Deployable = void 0;
const events_1 = require("events");
const cli_ux_1 = require("cli-ux");
const prompter_1 = require("./prompter");
class Deployable {
}
exports.Deployable = Deployable;
/**
 * Interface for deploying Salesforce Org metadata.
 */
class Deployer extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        /**
         * Deployables are individual pieces that can be deployed on their own. For example,
         * each pacakge in a salesforce project is considered a deployable that can be deployed
         * on its own.
         */
        this.deployables = [];
        this.prompter = new prompter_1.Prompter();
    }
    /**
     * Method for displaying deploy progress to the user
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
    progress(current, total, message) { }
    /**
     * Log messages to the console
     */
    log(msg, ...args) {
        cli_ux_1.default.log(msg, ...args);
    }
    /**
     * Prompt user for additional information
     */
    async prompt(questions, initialAnswers) {
        const answers = await this.prompter.prompt(questions, initialAnswers);
        return answers;
    }
    /**
     * Overwrite the deployables property on the class.
     */
    selectDeployables(deployables) {
        this.deployables = Object.assign([], deployables);
    }
}
exports.Deployer = Deployer;
