// salesforce-alm/dist/commands/force/org/delete.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrgDeleteCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const sandboxOrgConfig_1 = require("@salesforce/core/lib/config/sandboxOrgConfig");
const ts_types_1 = require("@salesforce/ts-types");
const sandboxOrg_1 = require("../../../lib/org/sandbox/sandboxOrg");
const activeScratchOrgDeleteApi_1 = require("../../../lib/org/activeScratchOrgDeleteApi");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('salesforce-alm', 'org_delete');
class OrgDeleteCommand extends command_1.SfdxCommand {
    async run() {
        let deleteMsgKey;
        const prodOrgUsername = await this.org.getSandboxOrgConfigField(sandboxOrgConfig_1.SandboxOrgConfig.Fields.PROD_ORG_USERNAME);
        if (prodOrgUsername) {
            deleteMsgKey = messages.getMessage('sandbox');
        }
        else {
            deleteMsgKey = messages.getMessage('scratchOrg');
        }
        // we either need permission to proceed without a prompt OR get the user to confirm
        if (this.flags.noprompt ||
            (await this.ux.confirm(messages.getMessage('confirmDeleteYesNo', [deleteMsgKey, this.org.getUsername()])))) {
            return prodOrgUsername ? this.deleteSandbox(prodOrgUsername) : this.deleteScratchOrg();
        }
    }
    async deleteSandbox(prodOrgUsername) {
        let successMessageKey = 'commandSandboxSuccess';
        this.logger.debug('Delete started for sandbox org %s ', this.org.getUsername());
        const prodOrg = await core_1.Org.create({
            aliasOrUsername: ts_types_1.ensureString(prodOrgUsername),
            aggregator: this.configAggregator,
        });
        const prodSandboxOrg = sandboxOrg_1.SandboxOrg.getInstance(prodOrg, this.flags.wait, this.logger, this.flags.clientid);
        try {
            await prodSandboxOrg.deleteSandbox(this.org.getOrgId());
            this.logger.debug('Sandbox org %s successfully marked for deletion', this.org.getUsername());
        }
        catch (err) {
            if (err.name === 'sandboxProcessNotFoundByOrgId') {
                successMessageKey = 'commandSandboxConfigOnlySuccess';
            }
            else {
                throw err;
            }
        }
        await this.org.remove();
        this.logger.debug('Sandbox org config %s has been successfully deleted', this.org.getUsername());
        this.ux.log(messages.getMessage(successMessageKey, [this.org.getUsername()]));
    }
    async deleteScratchOrg() {
        let alreadyDeleted = false;
        try {
            await activeScratchOrgDeleteApi_1.deleteScratchOrgFromDevHub(this.org, this.hubOrg);
        }
        catch (err) {
            if (err.name === 'attemptingToDeleteExpiredOrDeleted') {
                alreadyDeleted = true;
            }
            else {
                // Includes the "insufficientAccessToDelete" error.
                throw err;
            }
        }
        await this.org.remove();
        this.ux.log(messages.getMessage(alreadyDeleted ? 'deleteOrgConfigOnlyCommandSuccess' : 'deleteOrgCommandSuccess', [
            this.org.getUsername(),
        ]));
        return {
            orgId: this.org.getOrgId(),
            username: this.org.getUsername(),
        };
    }
}
exports.OrgDeleteCommand = OrgDeleteCommand;
OrgDeleteCommand.help = messages.getMessage('commandHelp');
/*
 * TODO: When SfdxCommand change the messages for description to include both the description and the help messages
 *       we have to remove the help from description.
 */
OrgDeleteCommand.description = messages.getMessage('commandDescription') + '\n\n' + OrgDeleteCommand.help;
OrgDeleteCommand.requiresUsername = true;
OrgDeleteCommand.supportsDevhubUsername = true;
OrgDeleteCommand.flagsConfig = {
    noprompt: command_1.flags.boolean({
        char: 'p',
        description: messages.getMessage('forceFlagDescription'),
    }),
};
